distchan.go

```diff
@@ -4,15 +4,41 @@ import (
 	"bytes"
 	"encoding/binary"
 	"encoding/gob"
+	"errors"
 	"io"
 	"log"
 	"net"
 	"os"
 	"reflect"
+	"runtime"
 	"sync"
-	"time"
+	"sync/atomic"
 )
 
+// ErrorInIsNotChannel raises when 'in' is not nil and not a channel
+// ErrorOutIsNotChannel raises when 'out' is not nil and not a channel
+var (
+	ErrorInIsNotChannel  = errors.New("Parameter 'in' is not a channel")
+	ErrorOutIsNotChannel = errors.New("Parameter 'out' is not a channel")
+	ErrorBadRequest      = errors.New("Bad request format")
+)
+
+// sync.Pool for connection buffers
+var bufPool = sync.Pool{}
```

Comment:
sync pool is safe for goroutines, reuses memory and protected from garbage collector. your buffers in []byte was garbage collected and not reuse memory.

Reply:
I don't believe the previous buffers would have been garbage-collected, because there was a constant reference to them through the server's `connections` slice. Do you have any additional reason to believe otherwise?

I do still like the pool idea, mostly for clarity and simplicity, so regardless of GC issues I think it might be worth switching.

Reply:
just trace cpu with pprof and see how GC is work on this slices in concurrent tests. i do it for another project and see it. sync.pool is very good for concurrent reusing memory.